nxthat/nanocl
# feat: save tcp/udp request inside database

**- What I did**
- Parse the TCP/UDP request and save them in the database.
- Create a [generic insert](https://github.com/Narayanbhat166/nanocl/blob/e39e9c80dc7100f13dfa4e835eb5f60937cbd1e1/bin/nanocld/src/repositories/http_metric.rs#L66) function to insert entities into the database. I could have created another function as written [here](https://github.com/Narayanbhat166/nanocl/blob/e39e9c80dc7100f13dfa4e835eb5f60937cbd1e1/bin/nanocld/src/repositories/http_metric.rs#L66) to insert the stream metric, but most of the logic would remain the same, so created a generic function which can be used to insert any Database entities. The http metric still uses the old function. Let me know if I have to use the generic function for inserting http metric too.

**- How I did it**
Create a struct called `StreamMetricPartial` and them implement `to_db_model` Trait. After converting it to the db model, save it in the database by calling the `generic_insert` function.

**- How to verify it**
- Deploy the `ProxyRule` as mentioned here https://github.com/nxthat/nanocl/issues/332#issuecomment-1605356260
- go to http://localhost:5555
- Check if there is an entry in the `stream_metrics` table.
<img width="764" alt="Screenshot 2023-07-04 at 10 51 50 PM" src="https://github.com/nxthat/nanocl/assets/48803246/19b13615-a5d8-439b-a079-593b9709ec84">

**- Description for the changelog**

closes #332 
feat: save tcp/udp request inside the database

**- A picture of a cute animal (not mandatory but encouraged)**
![cute_cat](https://github.com/nxthat/nanocl/assets/48803246/7fdfe4c4-b49e-4c7e-9943-a4e0a4f456df)
